std/algorithm/iteration.d

```diff
@@ -4289,3 +4289,409 @@ unittest
+A k-combination with repetitions, or k-multicombination, or multisubset of size k from a set S is given by a sequence of k not necessarily distinct elements of S, where order is not taken into account.
+Imagine this as the product filtered for only ordered items.
+
+For example $(D"AB".combinationsRepeat(2).array) returns $(D["AA", "AB", "BB"]).
+
+Params:
+    r = RandomAccessRange or string as origin
+    k = number of combinations
+
+Returns:
+    Forward range which yields the k-multicombinations items
+*/
+auto combinationsRepeat(Range)(Range r, size_t k=1)
+if (isRandomAccessRange!Range && hasLength!Range)
+in {
+    assert(k >= 1, "Invalid k");
+} body
+{
+    import std.array: array;
+    import std.range: indexed;
+    static struct CombinationRepeat{
+        ElementType!Range[] front;
+        private Range _r;
+        private size_t[] _state;
+
+        private size_t _repeat, _length;
+        private size_t _max_states, _pos;
+        private bool _lengthComputed;
+
+        bool empty;
+
+        this(Range r, size_t repeat) pure nothrow @safe
+        {
+            _r = r;
+            _length = r.length;
+            _repeat = repeat;
+
+            // set initial state
+            _max_states = length;
+            if (_length > 0 && _max_states > 0){
+                _state = new size_t[repeat];
+                front = _r.indexed(_state).array;
+            } else {
+                empty = true;
+            }
+        }
+
+        @property typeof(this) save()
+        {
+            auto c = typeof(this)(_r, _repeat);
+            c._state = _state.dup;
+            c._pos = _pos;
+            c.empty = empty;
+            return c;
+        }
+
+        void popFront() pure nothrow @safe {
+            if (empty) return;
+            // check whether we have reached the end
+            _pos++;
+            if (_pos == _max_states){
+                empty = true;
+                return;
+            }
+            //do _getNextState();  while(!_state.isSorted);
+            size_t i = _repeat - 1;
+            // go to next settable item (not at end state)
+            while (_state[i] == _length - 1) {
+                i--;
+            }
+            _state[i] = _state[i] + 1;
+            if (i == 0){
+                // reset all bits and start at the beginning
+                foreach_reverse (ref el; _state){
+                    el = _state[i];
+                }
+            }
+            front = _r.indexed(_state).array;
+        }
+
+        @property size_t length() pure nothrow @nogc {
+            if (!_lengthComputed){
+                if (_length == 0){
+                    _max_states = 0;
+                } else {
+                    _max_states = binomial(_length + _repeat  - 1, _repeat);
+                }
+                _lengthComputed = true;
+            }
+            return _max_states;
+        }
+    }
+    return CombinationRepeat(r, k);
+}
+
+///
+unittest
```

Comment:
since the range primitive are marked as `pure nothrow @safe` you should also mark the unittest so, in order to verify that they really are